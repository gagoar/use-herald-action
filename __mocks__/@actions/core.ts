export const setFailed = jest.fn();
export const getInput = jest.fn();
export const setOutput = jest.fn();
