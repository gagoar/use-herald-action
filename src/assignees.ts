/* eslint-disable @typescript-eslint/camelcase */
import { Octokit } from '@octokit/rest';
import { MatchingRule } from './rules';
import PQueue from 'p-queue';

export const handleAssignees = async (
  client: InstanceType<typeof Octokit>,
  owner: string,
  repo: string,
  prNumber: number,
  matchingRules: MatchingRule[],
  requestConcurrency = 1
) => {
  const queue = new PQueue({ concurrency: requestConcurrency });

  return Promise.all(
    matchingRules.map((matchingRule) =>
      queue.add(() =>
        client.issues.addAssignees({
          owner,
          repo,
          issue_number: prNumber,
          assignees: matchingRule.users,
        })
      )
    )
  );
};
