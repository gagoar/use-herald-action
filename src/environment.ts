import { join } from 'path';
import envalid, { str, testOnly } from 'envalid';

const environment = () =>
  envalid.cleanEnv(
    process.env,
    {
      GITHUB_EVENT_PATH: str({
        devDefault: testOnly('__mocks__/event.json'),
      }),
      GITHUB_WORKSPACE: str({
        devDefault: testOnly(join(__dirname, '..')),
      }),
      GITHUB_EVENT_NAME: str({ devDefault: 'pull_request' }),
      GITHUB_REPOSITORY: str({ devDefault: testOnly('someRepo') }),
      GITHUB_SHA: str({
        devDefault: testOnly('ffac537e6cbbf934b08745a378932722df287a53'),
      }),
      TASK_ID: str({ default: 'use-herald-action' }),
    },
    { dotEnvPath: null }
  );

export const env = environment();
