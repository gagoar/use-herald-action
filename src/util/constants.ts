export const maxPerPage = 100;
export const OUTPUT_NAME = 'appliedRules';
export const FILE_ENCODING = 'utf8';

export const EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
export enum SUPPORTED_EVENT_TYPES {
  PULL_REQUEST = 'pull_request',

  PULL_REQUEST_TARGET = 'pull_request_target',
  push = 'push',
}
interface Commit {
  sha: string;
}
interface Owner {
  login: string;
  id: number;
}
interface Repository {
  name: string;
  owner: Owner;
}
interface PullRequest {
  head: Commit;
  base: Commit;
  body: string;

  organization: string;
}
export interface Event {
  action: string;
  number: number;
  pull_request: PullRequest;

  repository: Repository;
}
