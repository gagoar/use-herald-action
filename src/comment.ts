import PQueue from 'p-queue';

import { composeCommentsForUsers } from './rules';

import { maxPerPage } from './util/constants';

import { Octokit, RestEndpointMethodTypes } from '@octokit/rest';
import { logger } from './util/debug';
import { ActionMapInput } from '.';
type AllCommentsParams = RestEndpointMethodTypes['issues']['listComments']['parameters'];

type AllCommentsResponse = RestEndpointMethodTypes['issues']['listComments']['response'];

type IssueComments = AllCommentsResponse['data'];

const debug = logger('comment');

const getAllComments = async (
  client: InstanceType<typeof Octokit>,
  params: AllCommentsParams
): Promise<IssueComments> => {
  const page = 1;
  const { data: comments } = await client.issues.listComments({
    ...params,
    per_page: maxPerPage,
    page,
  });

  if (comments.length < maxPerPage) {
    return comments;
  } else {
    const { data: moreComments } = await client.issues.listComments({
      ...params,
      page: page + 1,
      per_page: maxPerPage,
    });
    return [...comments, ...moreComments];
  }
};

export const handleComment: ActionMapInput = async (
  client,
  { owner, repo, prNumber, matchingRules },
  requestConcurrency = 1
): Promise<unknown> => {
  debug('handleComment called with:', matchingRules);

  const queue = new PQueue({ concurrency: requestConcurrency });
  const commentsFromRules = composeCommentsForUsers(matchingRules);
  const rawComments = await getAllComments(client, {
    owner,
    repo,
    issue_number: prNumber,
  });
  const comments = rawComments.map(({ body }) => body);

  const onlyNewComments = commentsFromRules.filter((comment: string) => !comments.includes(comment));

  debug('comments to add:', onlyNewComments);

  return Promise.all(
    onlyNewComments.map((body: string) => {
      return queue.add(() =>
        client.issues.createComment({
          owner,
          repo,
          issue_number: prNumber,
          body,
        })
      );
    })
  );
};
