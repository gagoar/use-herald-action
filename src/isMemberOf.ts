import { context } from '@actions/github';
import PQueue from 'p-queue';
import { catchHandler } from './util/catchHandler';
import { OctokitClient } from './util/constants';
import { logger } from './util/debug';

const debug = logger('isMemberOf');

const ACTIVE_STATE = 'active';

export const handleMembership = async (
  client: OctokitClient,
  isMemberOf: string[] = [],
  requestConcurrency = 2
): Promise<boolean> => {
  const { repo, actor } = context;

  const queue = new PQueue({ concurrency: requestConcurrency });

  const membershipChecks = isMemberOf.map((team) => {
    return {
      org: repo.owner,
      team_slug: team,
      username: actor,
    };
  });
  debug(
    `We will check membership of ${actor} in the following teams ${membershipChecks.map(({ team_slug: team }) => team)}`
  );

  const results = await Promise.all(
    membershipChecks.map((membership) => queue.add(() => client.teams.getMembershipForUserInOrg(membership)))
  ).catch(catchHandler(debug));

  const isMember = results instanceof Array && (results).some(result => result.data.state === ACTIVE_STATE);

  return isMember;
};
