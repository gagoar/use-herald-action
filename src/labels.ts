import { Octokit } from '@octokit/rest';
import { MatchingRule } from './rules';
import PQueue from 'p-queue';
import { logger } from './util/debug';
import { makeArray } from './util/makeArray';

const debug = logger('labels');

export const handleLabels = async (
  client: InstanceType<typeof Octokit>,
  owner: string,
  repo: string,
  prNumber: number,
  matchingRules: MatchingRule[],
  requestConcurrency = 1
): Promise<unknown> => {
  const queue = new PQueue({ concurrency: requestConcurrency });

  debug('called with:', matchingRules);

  const labels = matchingRules
    .filter(({ labels }) => labels)
    .reduce((memo, { labels }) => [...memo, ...makeArray(labels)], [] as string[]);

  debug('labels', labels);

  if (!labels.length) {
    debug('no labels where found');
    return undefined;
  }
  const result = await queue.add(() =>
    client.issues.addLabels({
      owner,
      repo,
      issue_number: prNumber,
      labels,
    })
  );

  debug('result:', result);
  return result;
};
