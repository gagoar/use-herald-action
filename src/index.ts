import { getInput, setOutput, setFailed } from '@actions/core';
import groupBy from 'lodash.groupby';
import { loadRules, getMatchingRules, RuleActions, allRequiredRulesHaveMatched, MatchingRule, Rule } from './rules';
import { Event, OUTPUT_NAME, SUPPORTED_EVENT_TYPES, RuleFile } from './util/constants';
import { logger } from './util/debug';
import { env } from './environment';

import { Octokit } from '@octokit/rest';
import { handleAssignees } from './assignees';
import { handleLabels } from './labels';
import { handleReviewers } from './reviewers';
import { handleStatus } from './statuses';
import { handleComment } from './comment';
import { loadJSONFile } from './util/loadJSONFile';
import { isEventSupported } from './util/isEventSupported';

const debug = logger('index');

const EnhancedOctokit = Octokit;

export enum Props {
  GITHUB_TOKEN = 'GITHUB_TOKEN',
  rulesLocation = 'rulesLocation',
  dryRun = 'dryRun',
  base = 'base',
}

export type ActionInput = {
  owner: string;
  repo: string;
  prNumber: number;
  matchingRules: MatchingRule[];
  rules: Rule[];
  sha: string;
  base: string;
  files: RuleFile[];
};
export type ActionMapInput = (
  client: InstanceType<typeof Octokit>,
  options: ActionInput,
  requestConcurrency?: number
) => Promise<unknown>;

const actionsMap: Record<RuleActions, ActionMapInput> = {
  [RuleActions.status]: handleStatus,
  [RuleActions.comment]: handleComment,
  [RuleActions.assign]: handleAssignees,
  [RuleActions.review]: handleReviewers,
  [RuleActions.label]: handleLabels,
};

type ActionName = keyof typeof RuleActions;

const getParams = () => {
  return Object.keys(Props).reduce((memo, prop) => {
    const value = getInput(prop);
    return value ? { ...memo, [prop]: value } : memo;
  }, {} as Partial<Record<keyof typeof Props, string>>);
};

export const main = async (): Promise<void> => {
  try {
    if (isEventSupported(env.GITHUB_EVENT_NAME)) {
      const event = loadJSONFile<Event>(env.GITHUB_EVENT_PATH);

      const {
        pull_request: {
          head: { sha: headSha },
          base: { sha: baseSha },
        },
        number: prNumber,
        repository: {
          name: repo,
          owner: { login: owner },
        },
      } = event;

      const { GITHUB_TOKEN, rulesLocation, base = baseSha, dryRun } = getParams();

      debug('params:', { rulesLocation, base, dryRun });

      if (!rulesLocation) {
        const message = `${Props.rulesLocation} is required`;
        setFailed(message);
        throw new Error(message);
      }

      const rules = loadRules(rulesLocation);

      debug('loaded rules and locations', {
        rules,
        dir: env.GITHUB_WORKSPACE,
        rulesLocation,
      });
      const client = new EnhancedOctokit({ auth: GITHUB_TOKEN });

      const {
        data: { files },
      } = await client.repos.compareCommits({
        base,
        head: headSha,
        owner,
        repo,
      });

      const matchingRules = getMatchingRules(
        rules,
        files,
        event,
        files.map(({ patch }) => patch)
      );

      debug('matchingRules:', matchingRules);

      if (!allRequiredRulesHaveMatched(rules, matchingRules)) {
        throw new Error(
          `Not all Rules with errorLevel set to error have matched. Please double check that these rules apply: ${rules
            .filter((rule) => rule.errorLevel && rule.errorLevel === 'error')
            .map((rule) => rule.name)
            .join(', ')}`
        );
      }

      const groupedRulesByAction = groupBy(matchingRules, (rule) => rule.action);

      if (dryRun !== 'true') {
        debug('not a dry Run');

        if (matchingRules.length) {
          const groupNames = Object.keys(groupedRulesByAction) as ActionName[];

          debug('groupNames', groupNames);

          await Promise.all([
            groupNames.map((actionName: ActionName) => {
              const action = actionsMap[RuleActions[actionName]];

              const options: ActionInput = {
                owner,
                repo,
                prNumber,
                matchingRules: groupedRulesByAction[RuleActions[actionName]],
                rules,
                sha: headSha,
                base: baseSha,
                files,
              };

              return action(client, options);
            }),
          ]);
        }
      }

      setOutput(OUTPUT_NAME, groupedRulesByAction);
    } else {
      setOutput(OUTPUT_NAME, []);
      throw new Error(
        `use-herald-action only supports [${Object.values(SUPPORTED_EVENT_TYPES).join(
          ', '
        )}] events for now, event found: ${env.GITHUB_EVENT_NAME}`
      );
    }
  } catch (e) {
    setFailed(e);
  }
};
