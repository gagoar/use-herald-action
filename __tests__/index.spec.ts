/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-var-requires */
import { Props } from '../src';
import { Event } from '../src/util/constants';
import * as actions from '@actions/core';
import { env } from '../src/environment';
import * as comment from '../src/comment';
import { mockCompareCommits } from './util/mockGitHubRequest';
import { Main, mockedInput } from './util/helpers';

jest.mock('@actions/core');
jest.mock('../src/comment');
jest.mock('../src/environment', () => {
  const { env } = jest.requireActual('../src/environment');

  return {
    env: {
      ...env,
      GITHUB_EVENT_PATH: '__mocks__/event.json',
      GITHUB_EVENT_NAME: 'pull_request',
    },
  };
});

const event = require(`../${env.GITHUB_EVENT_PATH}`) as Event;

const handleComment = comment.handleComment as jest.Mock<any>;
const setOutput = actions.setOutput as jest.Mock<any>;
const setFailed = actions.setFailed as jest.Mock<any>;
const getInput = actions.getInput as jest.Mock<any>;

const getGithubMock = () =>
  mockCompareCommits({
    login: event.repository.owner.login,
    name: event.repository.name,
    base: event.pull_request.base.sha,
    head: event.pull_request.head.sha,
  });

describe('use-herald-action', () => {
  beforeEach(() => {
    getInput.mockClear();
    setFailed.mockClear();
    setOutput.mockClear();
    handleComment.mockClear();
  });
  it('should fail if rulesLocation is not present', async () => {
    getInput.mockImplementation((key: Partial<keyof typeof mockedInput>) => {
      return key === Props.rulesLocation ? undefined : mockedInput[key];
    });

    const { main } = require('../src') as Main;

    await main();

    expect(setFailed).toHaveBeenCalled();
  });

  it('should fail if rules with errorLevel set to "error" does not match', async () => {
    const changedRulesDirectory = { ...mockedInput, [Props.rulesLocation]: '__mocks__/required_rules/*.json' };
    getInput.mockImplementation((key: Partial<keyof typeof changedRulesDirectory>) => {
      return changedRulesDirectory[key];
    });

    const github = getGithubMock();

    const { main } = require('../src') as Main;

    await main();

    expect(setFailed.mock.calls[0]).toMatchInlineSnapshot(`
      Array [
        "Not all Rules with errorLevel set to error have matched. Please double check that these rules apply: rule2.json, this rules should fail because the pattern is malformed",
      ]
    `);
    expect(setOutput).not.toHaveBeenCalled();
    expect(github.isDone()).toBe(true);
  });

  it('should run normally (with dryRun: true)', async () => {
    getInput.mockImplementation((key: Partial<keyof typeof mockedInput>) => {
      return mockedInput[key];
    });

    const github = getGithubMock();

    const { main } = require('../src') as Main;

    await main();

    expect(setFailed).not.toHaveBeenCalled();
    expect(setOutput).toHaveBeenCalled();
    expect(github.isDone()).toBe(true);
  });

  it('should run the entire action', async () => {
    const input = { ...mockedInput, [Props.dryRun]: 'false' };
    getInput.mockImplementation((key: Partial<keyof typeof input>) => {
      return input[key];
    });

    const github = getGithubMock();

    const { main } = require('../src') as Main;

    await main();

    expect(handleComment).toHaveBeenCalled();
    expect(setOutput.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "appliedRules",
          MatchingRules [
            Object {
              "action": "comment",
              "customMessage": "This is a custom message for a rule",
              "eventJsonPath": Array [],
              "excludes": Array [],
              "includes": Array [
                "*.ts",
              ],
              "includesInPatch": Array [],
              "matched": true,
              "name": "rule1.json",
              "path": "${env.GITHUB_WORKSPACE}/__mocks__/rules/rule1.json",
              "teams": Array [],
              "users": Array [
                "eeny",
                "meeny",
                "miny",
                "moe",
              ],
            },
          ],
        ],
      ]
    `);

    expect(github.isDone()).toBe(true);
  });

  it('should run the entire action (no rules found)', async () => {
    const input = {
      ...mockedInput,
      [Props.rulesLocation]: 'someWrongDirectory',
      [Props.dryRun]: false,
    };

    getInput.mockImplementation((key: Partial<keyof typeof mockedInput>) => {
      return input[key];
    });

    const github = getGithubMock();
    const { main } = require('../src') as Main;

    await main();

    expect(handleComment).not.toHaveBeenCalled();
    expect(setFailed.mock.calls).toMatchInlineSnapshot('Array []');
    expect(setOutput.mock.calls).toMatchSnapshot();
    expect(github.isDone()).toBe(true);
  });
});
