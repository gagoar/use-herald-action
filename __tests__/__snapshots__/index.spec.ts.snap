// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`use-herald-action should run the entire action (no rules found) 2`] = `
Array [
  Array [
    "appliedRules",
    MatchingRules [],
  ],
]
`;
